src/Avalonia.Animation/Animatable.cs

```diff
@@ -43,48 +49,195 @@ public Transitions Transitions
+                    AddTransitions(e.NewItems);
+                    break;
+                case NotifyCollectionChangedAction.Remove:
+                    RemoveTransitions(e.OldItems);
+                    break;
+                case NotifyCollectionChangedAction.Replace:
+                    RemoveTransitions(e.OldItems);
+                    AddTransitions(e.NewItems);
+                    break;
+                case NotifyCollectionChangedAction.Reset:
+                    throw new NotSupportedException("Transitions collection cannot be reset.");
             }
         }
 
-        protected override void OnPropertyChanged<T>(AvaloniaPropertyChangedEventArgs<T> change)
+        private void AddTransitions(IList items)
         {
-            if (_transitions is null || _previousTransitions is null || change.Priority == BindingPriority.Animation)
+            if (!_transitionsEnabled)
+            {
                 return;
+            }
 
-            // PERF-SENSITIVE: Called on every property change. Don't use LINQ here (too many allocations).
-            foreach (var transition in _transitions)
+            _transitionState ??= new Dictionary<ITransition, TransitionState>();
+
+            for (var i = 0; i < items.Count; ++i)
             {
-                if (transition.Property == change.Property)
+                var t = (ITransition)items[i];
+
+                _transitionState.Add(t, new TransitionState
                 {
-                    if (_previousTransitions.TryGetValue(change.Property, out var dispose))
-                        dispose.Dispose();
+                    BaseValue = GetAnimationBaseValue(t.Property),
+                });
+            }
+        }
 
-                    var instance = transition.Apply(
-                        this,
-                        Clock ?? Avalonia.Animation.Clock.GlobalClock,
-                        change.OldValue.GetValueOrDefault(),
-                        change.NewValue.GetValueOrDefault());
+        private void RemoveTransitions(IList items)
+        {
+            if (_transitionState is null)
+            {
+                return;
+            }
 
-                    _previousTransitions[change.Property] = instance;
-                    return;
+            for (var i = 0; i < items.Count; ++i)
+            {
+                var t = (ITransition)items[i];
+
+                if (_transitionState.TryGetValue(t, out var state))
+                {
+                    state.Instance?.Dispose();
+                    _transitionState.Remove(t);
                 }
             }
         }
+
+        private object GetAnimationBaseValue(AvaloniaProperty property)
+        {
+            var value = this.GetBaseValue(property, BindingPriority.LocalValue);
+
+            if (value == AvaloniaProperty.UnsetValue)
+            {
+                value = GetValue(property);
+            }
+
+            return value;
+        }
+
+        private class TransitionState
```

Comment:
Could be a struct?

Reply:
Yeah I did consider this, but because its `BaseValue` needs to be writable, and writable structs in a list aren't generally a great idea, I made it a class. Given that it's a private implementation detail though, it might not be so bad.